### DISCUSSION TITLE
Do you still use Gulp?

### DISCUSSION THREAD

**Discussion by 1337hero** [gulpjs/gulp#2725] [Polls]
I am really curious about this. Are people still using gulp in their workflows? What have you moved onto?

**Comment by angusmci** [comment]
Follow-up question: people who aren't using `gulp` any more, what are you using instead?

The responses I've seen to that question seem to break down into "use webpack" and "just use npm (or yarn)". But `webpack` and other bundlers seem to be much more specific and limited than `gulp` (and don't do the things that I want to do), while writing Javascript scripts and invoking them from `npm` seems to involve a great deal more work than using `gulp`.

Is there some better alternative to `gulp` out there now? Or are we just stuck in a situation where `gulp` is slowly dying, but there's no real replacement for it?

**Reply by yocontra** [reply]
Out of curiosity, what about gulp is "dying"? We've been doing 1.4M installs a week pretty consistently for the last decade, have been pretty much feature complete since v4, and there are a very small number of open issues. Fundamentally we don't believe in constantly pushing changes for the sake of change like some other tools where you need to rewrite your config every 6 months.

v5 is on the way with some improvements (mostly just updating things, and stream improvements) but nothing is fundamentally broken that needs fixing. If you're just looking at the main repo there hasn't been much git activity but if you look deeper at other repos theres regular work occurring from the team.

**Reply by angusmci** [reply]
First, I should apologize: it was rather rude of me to come into someone's GitHub discussions and suggest that their project might be moribund. Second, I should add that I'm personally a big fan of `gulp` and would like to use it for more and more projects.

That said, I think there is a _perception_ that `gulp` is in limbo. Trying to find information about the current state of `gulp`, I found a very limited amount of _new_ information about it, and a fair number of articles suggesting that task runners like `gulp` and `grunt` were old or outdated technology, and that bundlers like `webpack` or `parcel` were the way to go (as I said above, I don't agree with this view; I just see it expressed quite often).

In my own use of `gulp`, I've hit some dependency issues with supporting packages that cause deprecation warnings to appear when you install `gulp`. There are also questions about the compatibility of `gulp` with the newest versions of `node`. The version of `gulp` marked as 'latest' on GitHub is 4 years old, the Twitter feed is sparse (and recent posts are mostly about the difficulty of funding development), v5 is in the works but has no expected ship date, and so on. If others have had similar experiences, this may have contributed to the impression that many people seem to have that `gulp` is 'over'.

This is _not_ an attack on you or the other maintainers. I know how much work is involved in a big software project, and how hard it can be to keep an open-source project going when you, like everyone else, have other things to do. I'm not even saying that this impression is an accurate picture of how things are. I'm just saying that this is how it can _look_ from the outside.

I do think that the `gulp` team could do things to fight this perception: something on the website to say that v4.0.2 is the latest version but that v5 is coming might help make it clear that `gulp` is still very much alive. If the core `gulp` hasn't changed for a few years, maybe the website could highlight the packages that _have_ changed? More updates on Twitter (or Mastodon) might also help. And so on. I know that you guys would probably rather do coding than marketing -- but I think it might be helpful to invest a little time in raising the profile of `gulp` and showing that it's still useful and widely-used.

And obviously, this is also an area where the `gulp` community can and should help. Now that I know that v5 is still making headway, I have no excuse not to make a donation, so I'll get right on that. And maybe if we write posts about how we use `gulp`, how to do things in `gulp`, how to get started with `gulp`, and so on, that could help show the world that `gulp` is still relevant.

Anyway, thank you to you and the other maintainers. I'm very pleased to hear that `gulp` is still moving forward, and I will do what I can to help set the record straight on the status of `gulp`.

**Reply by AntonioRedondo** [reply]
Couldn't agree more with @angusmci, Gulp doesn't have a clear replacement. Gulp is purely a Node.js task executor that gives you total control to run tasks in series or parallel, and pipe the outputs. webpack and Parcel is a totally different philosophy, and uses the concept of entry points for specific file types. webpack and friends are good at frontend JavaScript. But as soon as you have a need outside of the standard flows defined it starts to be rigid and rely on plugins. Gulp excels at flexibility in processing raw text.

Best thing of all, Gulp doesn’t need replacement at all, not even new features, it already feels simple yet complete and powerful in v4. It just needs minimum maintenance to keep it up to date in its Node.js environment. Some brief update about v5 will help to calm anxious fellas. 😄

Kudos to its creator and maintainers! 💝

**Reply by Juicetan** [reply]
I still use gulp everyday and still love it.  What I'm starting to have to justify are its outdated dependencies, sub-dependencies, and community plugins that keep popping up on Dependabot vulnerability alerts.  I think _that_ is a sign that it is fading in popularity.

**Reply by 1337hero** [reply]
Yeah that's turning into a problem. 

I'm exploring using Node scripts. The times I use Gulp now are generally on static projects. But it's still useful.

**Comment by back-2-95** [comment]
We have a good bunch of projects using Gulp4 for sass to css and js minification. So we created Docker image where Gulp4 is stored in "stone". Developers and projects don't have anymore package.json etc to install Gulp4 toolchain. Instead they just run one docker command.

**Comment by Nicolas-Bouteille** [comment]
I have to say that I have had the exact same impression that @angusmci 
I had a lot of deprecated warnings when compiling sass, 
At first I thought it was because I was using the deprecated node-sass
But then I discovered some of them like chokidar 3 was because of Gulp
After more research on the web I was under the impression that gulp 4 had not been updated in 4 years and that the version 5 was probably not gonna make it any time soon, that the new way to go was npm scripts.
So I tested npm scripts and IMHO it is way much simpler and lighter
As of right now I have just finished refactoring my project not to use gulp anymore and none-sass, but npm scripts instead with dart-sass

**Reply by 1337hero** [reply]
Thanks for bringing this to my attention. Never considered doing this, despite the fact that almost all of Gulps plugins are wrappers.

**Reply by Nicolas-Bouteille** [reply]
Here's what it looks like :
```
"scripts": {
    "clean": "rm -f ./css/*",
    "compile_css_dev": "sass sass:css",
    "compile_css_prod": "sass --no-source-map sass:css",
    "post_process_dev": "postcss ./css/*.css --use postcss-combine-media-query --replace",
    "post_process_prod": "postcss ./css --use autoprefixer postcss-combine-media-query --replace",
    "css-dev": "npm run clean && npm run compile_css_dev && npm run post_process_dev",
    "css-prod": "npm run clean && npm run compile_css_prod && npm run post_process_prod"
  },
```
or in a separate bash file with a dynamic $path, you can use something like : 
```
rm -rf $path/css/*
npx sass --no-source-map $path/sass:$path/css
npx postcss $path/css/*.css --use autoprefixer postcss-combine-media-query --replace

**Comment by gregzawadzki** [comment]
I used grunt, and it was limited.
I seen webpack and didnt like it at all.
I discovered gulp and I am amazed and in love with it. Even tho I'm not that much of a node fan, since I don't do much and its harder for me to do node things. But i have set my gulp files and they work like a charm once set :) 
Finally i don't have to remember what command I used to compile sass . If it was compass or dart sass or node sass etc. Everything is done by gulp :)

**Comment by Artemsites** [comment]
Please tell me if gulp will be supported? Is it relevant to use it today on 8.11.2023?

**Reply by ajiho** [reply]
I still use Gulp a lot and I really like it

**Reply by ChetSocio** [reply]
What about severe vulnerabilities it shows when we run npm install gulp ?? Can we ignore it? I am junior and working on creating own npm packages and thought gulp will help me some way. 
```

npm i gulp@4.0.2
npm WARN deprecated chokidar@2.1.8: Chokidar 2 does not receive security updates since 2019. Upgrade to chokidar 3 with 15x fewer dependencies

added 158 packages, removed 76 packages, changed 28 packages, and audited 342 packages in 25s

16 packages are looking for funding
  run `npm fund` for details

6 high severity vulnerabilities

To address issues that do not require attention, run:
  npm audit fix

To address all issues (including breaking changes), run:
  npm audit fix --force

Run `npm audit` for details.

```
When I run `npm audit fix --force `  It install previous versions with more number of severe vulnerabilities :/

**Comment by ilyaigpetrov** [comment]
@ChetSocio, I wouldn't bother about it as long as gulp's input is controlled only by you or your team.
Then the only one who may DDOS or hack your service another way is yourself.

**Comment by UweKeim** [comment]
Releasing a version 5 that breaks "everything" and having no documentation for version 5, certainly helps in speeding up killing Gulp.

**Reply by 1337hero** [reply]
LOL, it's a TODO on the project: https://github.com/orgs/gulpjs/projects/5?pane=issue&itemId=58055624

I am sure it will get done eventually. Kind of bad practice though to do it after the fact and not as a task for a rollout.

**Reply by AntonioRedondo** [reply]
You can find an v5 announcing post on Medium:
https://medium.com/gulpjs/announcing-gulp-v5-c67d077dbdb7

That post contains a link to the list of breaking changes:
https://github.com/gulpjs/gulp/blob/master/CHANGELOG.md#500-2024-03-25

It looks like there are not that many breaking changes unless you used advanced options.

**Reply by UweKeim** [reply]
`src()` shredding _all_ of my binary files is hell of a breaking change in my opinion. (And [the new option to fix it](https://github.com/orgs/gulpjs/discussions/2791#discussioncomment-9070983) isn't even documented anywhere).

**Comment by 1337hero** [comment]
I have since moved on to other lands, I haven't had a project need gulp since I made this post. Mostly becuase I work with Vite in React projects now or end up doing Vite and Astro.

I don't even find myself doing SASS anymore with things like Tailwind.
